File: library/src/test/java/io/github/anderscheow/validator/rules/BaseRuleTest.java
package io.github.anderscheow.validator.rules;

import android.support.annotation.StringRes;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import io.github.anderscheow.validator.conditions.common.And;

import static org.junit.Assert.*;

public class BaseRuleTest {

    private BaseRule baseRule;

    @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
        baseRule = null;
    }

    @Test
    public void errorMessage_DefaultErrorMessage() throws Exception {
        String errorMessage = "Invalid input";

        baseRule = new BaseRule() {
            @Override
            public boolean validate(String value) {
                return false;
            }
        };

        assertEquals(errorMessage, baseRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorMessage() throws Exception {
        String errorMessage = "This is a custom error message";

        baseRule = new BaseRule(errorMessage) {
            @Override
            public boolean validate(String value) {
                return false;
            }
        };

        assertEquals(errorMessage, baseRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorRes() throws Exception {
        @StringRes int errorRes = 0;

        baseRule = new BaseRule(errorRes) {
            @Override
            public boolean validate(String value) {
                return false;
            }
        };

        assertEquals(errorRes, baseRule.errorRes());
    }
}

File: library/src/test/java/io/github/anderscheow/validator/rules/BaseRuleTest.kt
package io.github.anderscheow.validator.rules

import androidx.annotation.StringRes
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class BaseRuleTest {

    private lateinit var baseRule: BaseRule

    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_DefaultErrorMessage() {
        val errorMessage = "Invalid input"

        baseRule = object : BaseRule() {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(errorMessage, baseRule.getErrorMessage())
        assertTrue(baseRule.isErrorAvailable)
        assertTrue(baseRule.isErrorMessageAvailable)
        assertFalse(baseRule.isErrorResAvailable)
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorMessage() {
        val errorMessage = "This is a custom error message"

        baseRule = object : BaseRule(errorMessage) {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(errorMessage, baseRule.getErrorMessage())
        assertTrue(baseRule.isErrorAvailable)
        assertTrue(baseRule.isErrorMessageAvailable)
        assertFalse(baseRule.isErrorResAvailable)
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorRes() {
        @StringRes val errorRes = 0

        baseRule = object : BaseRule(errorRes) {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(errorRes, baseRule.getErrorRes())
        assertTrue(baseRule.isErrorAvailable)
        assertTrue(baseRule.isErrorMessageAvailable)
        assertTrue(baseRule.isErrorResAvailable)
    }
}